modules/cudaimgproc/src/cuda/moments.cu

```diff
@@ -0,0 +1,137 @@
+// of this distribution and at http://opencv.org/license.html.
+
+#if !defined CUDA_DISABLER
+
+#include "opencv2/core/cuda/common.hpp"
+#include "opencv2/core/cuda/emulation.hpp"
+#include "opencv2/core/cuda/transform.hpp"
+#include "opencv2/core/cuda/functional.hpp"
+#include "opencv2/core/cuda/utility.hpp"
+#include "opencv2/core/cuda.hpp"
+
+using namespace cv::cuda;
+using namespace cv::cuda::device;
+
+
+namespace cv { namespace cuda { namespace device { namespace imgproc {
+
+constexpr int blockSizeX = 16;
+constexpr int blockSizeY = 16;
+constexpr int momentsSize = sizeof(cv::Moments) / sizeof(double);
+
+constexpr int m00 = offsetof(cv::Moments, m00) / sizeof(double);
+constexpr int m10 = offsetof(cv::Moments, m10) / sizeof(double);
+constexpr int m01 = offsetof(cv::Moments, m01) / sizeof(double);
+constexpr int m20 = offsetof(cv::Moments, m20) / sizeof(double);
+constexpr int m11 = offsetof(cv::Moments, m11) / sizeof(double);
+constexpr int m02 = offsetof(cv::Moments, m02) / sizeof(double);
+constexpr int m30 = offsetof(cv::Moments, m30) / sizeof(double);
+constexpr int m21 = offsetof(cv::Moments, m21) / sizeof(double);
+constexpr int m12 = offsetof(cv::Moments, m12) / sizeof(double);
+constexpr int m03 = offsetof(cv::Moments, m03) / sizeof(double);
+
+constexpr int mu20 = offsetof(cv::Moments, mu20) / sizeof(double);
+constexpr int mu11 = offsetof(cv::Moments, mu11) / sizeof(double);
+constexpr int mu02 = offsetof(cv::Moments, mu02) / sizeof(double);
+constexpr int mu30 = offsetof(cv::Moments, mu30) / sizeof(double);
+constexpr int mu21 = offsetof(cv::Moments, mu21) / sizeof(double);
+constexpr int mu12 = offsetof(cv::Moments, mu12) / sizeof(double);
+constexpr int mu03 = offsetof(cv::Moments, mu03) / sizeof(double);
+
+__global__ void ComputeSpatialMoments(const cuda::PtrStepSzb img, bool binary, double* moments) {
+    const unsigned int x = blockIdx.x * blockDim.x + threadIdx.x;
+    const unsigned int y = blockIdx.y * blockDim.y + threadIdx.y;
+    if (y < img.rows && x < img.cols) {
+      const unsigned int img_index = y * img.step + x;
+      const unsigned char val = (!binary || img.data[img_index] == 0) ? img.data[img_index] : 1;
+      if (val > 0) {
+        const unsigned long x2 = x * x, x3 = x2 * x;
+        const unsigned long y2 = y * y, y3 = y2 * y;
+
+        atomicAdd(&moments[m00],           val);
+        atomicAdd(&moments[m10], x       * val);
+        atomicAdd(&moments[m01],      y  * val);
+        atomicAdd(&moments[m20], x2      * val);
+        atomicAdd(&moments[m11], x  * y  * val);
+        atomicAdd(&moments[m02],      y2 * val);
+        atomicAdd(&moments[m30], x3      * val);
+        atomicAdd(&moments[m21], x2 * y  * val);
+        atomicAdd(&moments[m12], x  * y2 * val);
+        atomicAdd(&moments[m03],      y3 * val);
+      }
+    }
+}
+
+__global__ void ComputeCenteroid(const double* moments, double2* centroid) {
+    centroid->x = moments[m10] / moments[m00];
```

Comment:
It may be faster to calculate this every time inside `ComputeCenteralMoments` rather than launching a seperate kernel.